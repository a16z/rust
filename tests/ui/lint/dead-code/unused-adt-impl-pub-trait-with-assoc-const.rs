#![deny(dead_code)]

struct T1; //~ ERROR struct `T1` is never constructed
struct T2; //~ ERROR struct `T2` is never constructed
pub struct T3(i32); //~ ERROR struct `T3` is never constructed
pub struct T4(i32); //~ ERROR field `0` is never read

trait Trait1 { //~ ERROR trait `Trait1` is never used
    const UNUSED: i32;
    fn unused(&self) {}
    fn construct_self() -> Self;
}

pub trait Trait2 {
    const MAY_USED: i32;
    fn may_used(&self) {}
}

pub trait Trait3 {
    const MAY_USED: i32;
    fn may_used() -> Self;
}

impl Trait1 for T1 {
    const UNUSED: i32 = 0;
    fn construct_self() -> Self {
        Self
    }
}

impl Trait1 for T2 {
    const UNUSED: i32 = 0;
    fn construct_self() -> Self {
        Self
    }
}

impl Trait2 for T1 {
    const MAY_USED: i32 = 0;
}

impl Trait2 for T2 {
    const MAY_USED: i32 = 0;
}

impl Trait2 for T3 {
    const MAY_USED: i32 = 0;
}

impl Trait3 for T2 {
    const MAY_USED: i32 = 0;
    fn may_used() -> Self {
        Self
    }
}

impl Trait3 for T4 {
    const MAY_USED: i32 = 0;
    fn may_used() -> Self {
        T4(0)
    }
}

fn main() {}
