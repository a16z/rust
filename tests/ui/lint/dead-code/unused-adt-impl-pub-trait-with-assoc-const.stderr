error: struct `T1` is never constructed
  --> $DIR/unused-adt-impl-pub-trait-with-assoc-const.rs:3:8
   |
LL | struct T1;
   |        ^^
   |
note: the lint level is defined here
  --> $DIR/unused-adt-impl-pub-trait-with-assoc-const.rs:1:9
   |
LL | #![deny(dead_code)]
   |         ^^^^^^^^^

error: struct `T2` is never constructed
  --> $DIR/unused-adt-impl-pub-trait-with-assoc-const.rs:4:8
   |
LL | struct T2;
   |        ^^

error: struct `T3` is never constructed
  --> $DIR/unused-adt-impl-pub-trait-with-assoc-const.rs:5:12
   |
LL | pub struct T3(i32);
   |            ^^

error: field `0` is never read
  --> $DIR/unused-adt-impl-pub-trait-with-assoc-const.rs:6:15
   |
LL | pub struct T4(i32);
   |            -- ^^^
   |            |
   |            field in this struct
   |
   = help: consider removing this field

error: trait `Trait1` is never used
  --> $DIR/unused-adt-impl-pub-trait-with-assoc-const.rs:8:7
   |
LL | trait Trait1 {
   |       ^^^^^^

error: aborting due to 5 previous errors

